File: viskores/filter/entity_extraction/testing/UnitTestExternalFacesFilter.cxx
//============================================================================
//  Copyright (c) Kitware, Inc.
//  All rights reserved.
//  See LICENSE.txt for details.
//
//  This software is distributed WITHOUT ANY WARRANTY; without even
//  the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
//  PURPOSE.  See the above copyright notice for more information.
//============================================================================

#include <viskores/cont/testing/MakeTestDataSet.h>
#include <viskores/cont/testing/Testing.h>

#include <viskores/filter/clean_grid/CleanGrid.h>
#include <viskores/filter/entity_extraction/ExternalFaces.h>

using viskores::cont::testing::MakeTestDataSet;

namespace
{

// convert a 5x5x5 uniform grid to unstructured grid
viskores::cont::DataSet MakeDataTestSet1()
{
  viskores::cont::DataSet ds = MakeTestDataSet().Make3DUniformDataSet1();

  viskores::filter::clean_grid::CleanGrid clean;
  clean.SetCompactPointFields(false);
  clean.SetMergePoints(false);
  return clean.Execute(ds);
}

viskores::cont::DataSet MakeDataTestSet2()
{
  return MakeTestDataSet().Make3DExplicitDataSet5();
}

viskores::cont::DataSet MakeDataTestSet3()
{
  return MakeTestDataSet().Make3DUniformDataSet1();
}

viskores::cont::DataSet MakeDataTestSet4()
{
  return MakeTestDataSet().Make3DRectilinearDataSet0();
}

viskores::cont::DataSet MakeDataTestSet5()
{
  return MakeTestDataSet().Make3DExplicitDataSet6();
}

void TestExternalFacesExplicitGrid(const viskores::cont::DataSet& ds,
                                   bool compactPoints,
                                   viskores::Id numExpectedExtFaces,
                                   viskores::Id numExpectedPoints = 0,
                                   bool passPolyData = true)
{
  //Run the External Faces filter
  viskores::filter::entity_extraction::ExternalFaces externalFaces;
  externalFaces.SetCompactPoints(compactPoints);
  externalFaces.SetPassPolyData(passPolyData);
  viskores::cont::DataSet resultds = externalFaces.Execute(ds);

  // verify cellset
  viskores::cont::CellSetExplicit<> new_cellSet =
    resultds.GetCellSet().AsCellSet<viskores::cont::CellSetExplicit<>>();
  const viskores::Id numOutputExtFaces = new_cellSet.GetNumberOfCells();
  VISKORES_TEST_ASSERT(numOutputExtFaces == numExpectedExtFaces, "Number of External Faces mismatch");

  // verify fields
  VISKORES_TEST_ASSERT(resultds.HasField("pointvar"), "Point field not mapped successfully");
  VISKORES_TEST_ASSERT(resultds.HasField("cellvar"), "Cell field not mapped successfully");

  // verify CompactPoints
  if (compactPoints)
  {
    viskores::Id numOutputPoints = resultds.GetCoordinateSystem(0).GetNumberOfPoints();
    VISKORES_TEST_ASSERT(numOutputPoints == numExpectedPoints,
                     "Incorrect number of points after compacting");
  }
}

void TestWithHexahedraMesh()
{
  std::cout << "Testing with Hexahedra mesh\n";
  viskores::cont::DataSet ds = MakeDataTestSet1();
  std::cout << "Compact Points Off\n";
  TestExternalFacesExplicitGrid(ds, false, 96); // 4x4 * 6 = 96
  std::cout << "Compact Points On\n";
  TestExternalFacesExplicitGrid(ds, true, 96, 98); // 5x5x5 - 3x3x3 = 98
}

void TestWithHeterogeneousMesh()
{
  std::cout << "Testing with Heterogeneous mesh\n";
  viskores::cont::DataSet ds = MakeDataTestSet2();
  std::cout << "Compact Points Off\n";
  TestExternalFacesExplicitGrid(ds, false, 12);
  std::cout << "Compact Points On\n";
  TestExternalFacesExplicitGrid(ds, true, 12, 11);
}

void TestWithUniformMesh()
{
  std::cout << "Testing with Uniform mesh\n";
  viskores::cont::DataSet ds = MakeDataTestSet3();
  std::cout << "Compact Points Off\n";
  TestExternalFacesExplicitGrid(ds, false, 16 * 6);
  std::cout << "Compact Points On\n";
  TestExternalFacesExplicitGrid(ds, true, 16 * 6, 98);
}

void TestWithRectilinearMesh()
{
  std::cout << "Testing with Rectilinear mesh\n";
  viskores::cont::DataSet ds = MakeDataTestSet4();
  std::cout << "Compact Points Off\n";
  TestExternalFacesExplicitGrid(ds, false, 16);
  std::cout << "Compact Points On\n";
  TestExternalFacesExplicitGrid(ds, true, 16, 18);
}

void TestWithMixed2Dand3DMesh()
{
  std::cout << "Testing with mixed poly data and 3D mesh\n";
  viskores::cont::DataSet ds = MakeDataTestSet5();
  std::cout << "Compact Points Off, Pass Poly Data On\n";
  TestExternalFacesExplicitGrid(ds, false, 12);
  std::cout << "Compact Points On, Pass Poly Data On\n";
  TestExternalFacesExplicitGrid(ds, true, 12, 8);
  std::cout << "Compact Points Off, Pass Poly Data Off\n";
  TestExternalFacesExplicitGrid(ds, false, 6, 8, false);
  std::cout << "Compact Points On, Pass Poly Data Off\n";
  TestExternalFacesExplicitGrid(ds, true, 6, 5, false);
}

void TestExternalFacesFilter()
{
  TestWithHeterogeneousMesh();
  TestWithHexahedraMesh();
  TestWithUniformMesh();
  TestWithRectilinearMesh();
  TestWithMixed2Dand3DMesh();
}

} // anonymous namespace

int UnitTestExternalFacesFilter(int argc, char* argv[])
{
  return viskores::cont::testing::Testing::Run(TestExternalFacesFilter, argc, argv);
}

File: python_bindings/testing/unit_test_external_faces_filter.py
##=============================================================================
##
##  The contents of this file are covered by the Viskores license. See
##  LICENSE.txt for details.
##
##  By contributing to this file, all contributors agree to the Developer
##  Certificate of Origin Version 1.1 (DCO 1.1) as stated in DCO.txt.
##
##=============================================================================

from viskores.filter.clean_grid import CleanGrid
from viskores.filter.entity_extraction import ExternalFaces
from viskores.testing import MakeTestDataSet


def make_data_test_set1():
    dataset = MakeTestDataSet().Make3DUniformDataSet1()
    clean = CleanGrid()
    clean.SetCompactPointFields(False)
    clean.SetMergePoints(False)
    return clean.Execute(dataset)


def check_external_faces(dataset, compact_points, expected_cells, expected_points=None, pass_poly_data=True):
    external_faces = ExternalFaces()
    external_faces.SetCompactPoints(compact_points)
    external_faces.SetPassPolyData(pass_poly_data)
    result = external_faces.Execute(dataset)

    assert result.GetNumberOfCells() == expected_cells
    assert result.HasField("pointvar")
    assert result.HasField("cellvar")

    if compact_points and expected_points is not None:
        assert result.GetNumberOfPoints() == expected_points


def test_hexahedra_mesh():
    dataset = make_data_test_set1()
    check_external_faces(dataset, False, 96)
    check_external_faces(dataset, True, 96, 98)


def test_heterogeneous_mesh():
    dataset = MakeTestDataSet().Make3DExplicitDataSet5()
    check_external_faces(dataset, False, 12)
    check_external_faces(dataset, True, 12, 11)


def test_uniform_mesh():
    dataset = MakeTestDataSet().Make3DUniformDataSet1()
    check_external_faces(dataset, False, 16 * 6)
    check_external_faces(dataset, True, 16 * 6, 98)


def test_rectilinear_mesh():
    dataset = MakeTestDataSet().Make3DRectilinearDataSet0()
    check_external_faces(dataset, False, 16)
    check_external_faces(dataset, True, 16, 18)


def test_mixed_2d_and_3d_mesh():
    dataset = MakeTestDataSet().Make3DExplicitDataSet6()
    check_external_faces(dataset, False, 12)
    check_external_faces(dataset, True, 12, 8)
    check_external_faces(dataset, False, 6, pass_poly_data=False)
    check_external_faces(dataset, True, 6, 5, pass_poly_data=False)


def main():
    test_hexahedra_mesh()
    test_heterogeneous_mesh()
    test_uniform_mesh()
    test_rectilinear_mesh()
    test_mixed_2d_and_3d_mesh()


if __name__ == "__main__":
    main()
